Modules/_testcapi/pyrun.c

```diff
@@ -0,0 +1,92 @@
+#include "parts.h"
+#include "util.h"
+
+#include <stdio.h>
+#include <errno.h>
+
+
+static PyObject *
+run_fileexflags(PyObject *mod, PyObject *pos_args)
+{
+    PyObject *result = NULL;
+    const char *filename = NULL;
+    int start;
+    PyObject *globals = NULL;
+    PyObject *locals = NULL;
+    int closeit = 1;
+    PyCompilerFlags flags = _PyCompilerFlags_INIT;
+    PyCompilerFlags *pflags = NULL;
+    int cf_flags = 0;
+    int cf_feature_version = 0;
+    int fn;
+
+    FILE *fp = NULL;
+
+    if (!PyArg_ParseTuple(pos_args,
+                        "ziO|Oiii:eval_pyrun_fileexflags",
+                        &filename,
+                        &start,
+                        &globals,
+                        &locals,
+                        &closeit,
+                        &cf_flags,
+                        &cf_feature_version)) {
+        goto exit;
+    }
+
+    NULLABLE(globals);
+    NULLABLE(locals);
+    if (cf_flags || cf_feature_version) {
+        flags.cf_flags = cf_flags;
+        flags.cf_feature_version = cf_feature_version;
+        pflags = &flags;
+    }
+
+    if (globals && !PyDict_Check(globals)) {
+        PyErr_SetString(PyExc_TypeError, "globals must be a dict");
+        goto exit;
+    }
+
+    if (locals && !PyMapping_Check(locals)) {
+        PyErr_SetString(PyExc_TypeError, "locals must be a mapping");
+        goto exit;
+    }
+
+    fp = fopen(filename, "r");
```

Comment:
What if `filename` is NULL?

Reply:
Linux will return something like EFAULT with fp = NULL.
And it is not NULLABLE. Anyway, filename = NULL made test fail:

```
OSError: [Errno 14] Bad address
```
But NULL can be passed, if you use fp of stdin for example